### PR TITLE
Fixed issue #18699: LSActiveRecord::getMaxId() doesn't get refreshed when using from SurveyDynamic with multiple surveys on the same HTTP request

### PR DESCRIPTION
**Pull request by Shnoulle**
Dev: fix TokenDynamic and others too

### CODE REVIEW

**Comment by Shnoulle** [comment]
Need minIds too and reset to master

Just tell me if you think it's OK like that.

File: application/models/LSActiveRecord.php [hunk 1]
```diff
@@ -119,8 +122,17 @@ public function findAllAsArray($condition = '', $params = [])
      */
     public function getMaxId($field = null, $forceRefresh = false)
     {
-        static $maxIds = [];
-
+        $dynamicId = 0;
+        if (is_subclass_of($this, 'Dynamic')) {
+            /* \Response and \Token */
+            $dynamicId = $this->getDynamicId();
+        } elseif (!empty(self::$sid)) {
```
Comment (Shnoulle) [inline application/models/LSActiveRecord.php:131]: Must fix here, still have issue

File: application/models/TokenDynamic.php [hunk 2]
```diff
@@ -69,6 +69,7 @@ public static function model($sid = null)
         //We need to refresh if we changed sid
         if ($refresh === true) {
             $model->refreshMetaData();
+            $model::$maxIds = [];
```
[the remark this answers is not in the document]
Comment (Shnoulle) [reply application/models/TokenDynamic.php:72]: Yes :it's not OK here …

File: application/models/LSActiveRecord.php [hunk 3]
```diff
@@ -119,8 +124,17 @@ public function findAllAsArray($condition = '', $params = [])
      */
     public function getMaxId($field = null, $forceRefresh = false)
     {
-        static $maxIds = [];
-
+        $dynamicId = 0;
+        if (is_subclass_of($this, 'Dynamic')) {
+            /* \Response and \Token */
+            $dynamicId = $this->getDynamicId();
```
[the remark this answers is not in the document]
Comment (Shnoulle) [reply application/models/LSActiveRecord.php:130]: Yep

**Comment by Shnoulle** [comment]
> Besides the comments made, I still think we need to overload LSActiveRecord::RefreshMetadata() as to get maxIds() cleared.

Like put in mantis  comment : we need better static and cache system. here it's tracking down where there are potential issue.

**Comment by Shnoulle** [comment]
manual test done with

```
Question::model()->getMaxId();
QuestionGroup::model()->getMaxId();
Question::model()->getMinId();
QuestionGroup::model()->getMinId();
SurveyDynamic::model(131385)->getMaxId();
SurveyDynamic::model(918398)->getMaxId();
SurveyDynamic::model(131385)->getMinId();
SurveyDynamic::model(918398)->getMinId();
Response::model(131385)->getMaxId();
Response::model(918398)->getMaxId();
Response::model(131385)->getMinId();
Response::model(918398)->getMinId();
```

Question and QuestionGroup don't share maxIds

**Review by gabrieljenik — COMMENTED** [review]
Code looks OK.
I suggest a revision by Olle as well.
I believe a test of this whole mechanusm is needed before commiting.

Very similar to the tests you wrote on the comments but on a script

**Comment by Shnoulle** [comment]
> Very similar to the tests you wrote on the comments but on a script

Only with SurveyDynamic and Response , i can not know te number of Question and QuestionGroup, just it's different .

Maybe

`Question::model()->getMaxId() != QuestionGroup::model()->getMaxId()` ? If we import 2 survey before : it's not sure at 100% but at 99.9% ?

**Comment by gabrieljenik** [comment]
You can run a count on the DB...

If you prefer, I can handle it.
Let me know.

**Comment by Shnoulle** [comment]
> You can run a count on the DB...

You mean without using function :) start by 
1. count without function 
2. count with function : compare
3. count again with function : compare (check not same static value)

right ?

Why not : good idea.

**Review by gabrieljenik — APPROVED** [review]
Not tested. Will work on an automated test

**Comment by Shnoulle** [comment]
> Not tested. Will work on an automated test

need 5.X + conflict + master …

**Comment by Shnoulle** [comment]
master https://github.com/LimeSurvey/LimeSurvey/pull/3209

**Pull request closed without merging**